File: frontend/lib/api/metas.ts
import { del, get, patch, post, qs } from "./client"
import type {
  Meta,
  MetaContribution,
  MetaCreate,
  MetaPreview,
  MetaStatus,
  MonthSplit,
} from "./types"

export const listMetas = (month: string) => get<MetaStatus[]>(`/metas${qs({ month })}`)
export const listArchived = () => get<Meta[]>("/metas/archived")
export const monthSplit = (month: string) => get<MonthSplit>(`/metas/split${qs({ month })}`)
export const previewMeta = (month: string, body: MetaCreate) =>
  post<MetaPreview>(`/metas/preview${qs({ month })}`, body)
export const createMeta = (month: string, body: MetaCreate) =>
  post<MetaStatus>(`/metas${qs({ month })}`, body)
export const setMeta = (id: number, month: string, body: Partial<MetaCreate>) =>
  patch<MetaStatus>(`/metas/${id}${qs({ month })}`, body)
export const contribute = (id: number, month: string, amount: number) =>
  post<MetaContribution>(`/metas/${id}/contributions${qs({ month })}`, { amount })
export const cancelMeta = (id: number, month: string) => del<void>(`/metas/${id}${qs({ month })}`)
export const closeMeta = (id: number, month: string) =>
  post<MetaStatus>(`/metas/${id}/close${qs({ month })}`, {})
export const restoreMeta = (id: number, month: string) =>
  post<MetaStatus>(`/metas/${id}/restore${qs({ month })}`, {})

File: backend/src/quaestor/api/routers/metas.py
"""Metas REST router — thin adapter over services.metas.

Every refusal a meta can meet lives in the service, so this router carries no
rule of its own: it translates a request into a service call and the result
into JSON (ADR-0046).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlmodel import Session

from ...services import metas
from ...services import month as month_service
from ..deps import get_session
from ..schemas import (
    MetaContributionIn,
    MetaContributionOut,
    MetaCreate,
    MetaOut,
    MetaPreviewOut,
    MetaStatusOut,
    MetaUpdate,
    MonthSplitOut,
)

router = APIRouter(prefix="/metas", tags=["metas"])


@router.get("", response_model=list[MetaStatusOut])
def list_metas(month: str, session: Session = Depends(get_session)):
    return metas.list_metas(session, month)


@router.get("/split", response_model=MonthSplitOut)
def month_split(month: str, session: Session = Depends(get_session)):
    return month_service.split(session, month)


@router.get("/archived", response_model=list[MetaOut])
def list_archived(session: Session = Depends(get_session)):
    return metas.list_archived(session)


@router.post("/preview", response_model=MetaPreviewOut)
def preview_meta(body: MetaCreate, month: str, session: Session = Depends(get_session)):
    return metas.preview_meta(
        amount=body.amount,
        target_month=body.target_month,
        today=month,
        income=month_service.income_of(session, month),
    )


@router.post("", response_model=MetaOut, status_code=201)
def create_meta(body: MetaCreate, month: str, session: Session = Depends(get_session)):
    return metas.create_meta(session, today=month, **body.model_dump(exclude_unset=True))


@router.patch("/{meta_id}", response_model=MetaOut)
def set_meta(meta_id: int, body: MetaUpdate, month: str, session: Session = Depends(get_session)):
    return metas.set_meta(session, meta_id, today=month, **body.model_dump(exclude_unset=True))


@router.post("/{meta_id}/contributions", response_model=MetaContributionOut, status_code=201)
def contribute(meta_id: int, body: MetaContributionIn, month: str, session: Session = Depends(get_session)):
    metas.contribute(session, meta_id, year_month=month, amount=body.amount)
    return metas.contributions_of(session, meta_id)[-1]


@router.get("/{meta_id}/contributions", response_model=list[MetaContributionOut])
def list_contributions(meta_id: int, session: Session = Depends(get_session)):
    return metas.contributions_of(session, meta_id)


@router.delete("/contributions/{contribution_id}", status_code=204)
def remove_contribution(contribution_id: int, session: Session = Depends(get_session)):
    metas.remove_contribution(session, contribution_id)


@router.delete("/{meta_id}", status_code=204)
def cancel_meta(meta_id: int, month: str, session: Session = Depends(get_session)):
    metas.cancel_meta(session, meta_id, year_month=month)


@router.post("/{meta_id}/close", response_model=MetaOut)
def close_meta(meta_id: int, month: str, session: Session = Depends(get_session)):
    return metas.close_meta(session, meta_id, year_month=month)


@router.post("/{meta_id}/restore", response_model=MetaOut)
def restore_meta(meta_id: int, month: str, session: Session = Depends(get_session)):
    return metas.restore_meta(session, meta_id, today=month)
